### PR TITLE
Do not create dagruns for DAGs with import errors 

### PR DESCRIPTION
**Pull request by ephraimbuddy**
An active dag can suddenly have import errors as a result of the DAG file being changed. Currently, we
do not consider this before creating dagruns.

This PR adds the has_import_errors to `dag_model` so dags with import errors are not set to the scheduler to
create dagruns. Also, dag runs can no longer be created on the UI/API when the dag has import errors

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Review by jedcunningham — DISMISSED** [review]

**Review by kaxil — COMMENTED** [review]
Can we add a test for this case please?

Separately let's check why `max_active_runs` is not set via tests too

**Comment by ephraimbuddy** [comment]
> Can we add a test for this case please?
> 
> Separately let's check why `max_active_runs` is not set via tests too

I can't figure a test to add. 
My guess for the issue is that they removed `max_active_runs_per_dag`  from their configuration. 
To check this, I have added a fallback and also updated dag's test to ensure once DagModel is instantiated, it'll have a `max_active_runs`

File: airflow/models/dag.py [hunk 1]
```diff
@@ -2743,7 +2743,9 @@ def __init__(self, concurrency=None, **kwargs):
             else:
                 self.max_active_tasks = conf.getint('core', 'max_active_tasks_per_dag')
         if self.max_active_runs is None:
-            self.max_active_runs = conf.getint('core', 'max_active_runs_per_dag')
+            self.max_active_runs = kwargs.get(
+                'max_active_runs', conf.getint('core', 'max_active_runs_per_dag', fallback=16)
+            )
```
Comment (jedcunningham) [inline airflow/models/dag.py]: Wait, why do we need the fallback? It should get it from the default config anyways.

Comment (ephraimbuddy) [reply airflow/models/dag.py]: I just added it to be safe in case max_active_runs_per_dag is removed from the configuration. 
Can't think of any other reason why they'd get that errror

Comment (jedcunningham) [reply airflow/models/dag.py]: Gotcha. They'd have to remove it from the `config_templates/default_airflow.cfg` in their install, so I think that is unlikely to be the cause.

**Comment by ephraimbuddy** [comment]
I discovered that this happens because the DAG has import error and is active after upgrading. So scheduler was trying to create dagruns for it. 
I have updated the code and commit message. 
cc: @jedcunningham  @kaxil

File: airflow/models/dag.py [hunk 2]
```diff
@@ -2880,12 +2883,16 @@ def dags_needing_dagruns(cls, session: Session):
         # TODO[HA]: Bake this query, it is run _A lot_
         # We limit so that _one_ scheduler doesn't try to do all the creation
         # of dag runs
+        # exclude dags with import error
+        import_errors = [c.filename for c in session.query(importerror.filename).all()]
```
Comment (kaxil) [inline airflow/models/dag.py]: Can you test this with zip file too please

File: airflow/models/dag.py [hunk 3]
```diff
@@ -2880,12 +2883,16 @@ def dags_needing_dagruns(cls, session: Session):
...
         query = (
             session.query(cls)
             .filter(
+                cls.max_active_runs != None,  # noqa
```
Comment (ashb) [inline airflow/models/dag.py]: What does this have to do with import errors?

File: airflow/models/dag.py [hunk 4]
```diff
@@ -2880,12 +2883,16 @@ def dags_needing_dagruns(cls, session: Session):
...
                 cls.is_paused == expression.false(),
                 cls.is_active == expression.true(),
                 cls.next_dagrun_create_after <= func.now(),
+                cls.fileloc.notin_(import_errors),
```
Comment (ashb) [inline airflow/models/dag.py]: We should find another way of doing this -- this function is part of the "hot" path of the scheduler, and a) adding an extra query and b) sending (possibly) a large number of params via this "NOT IN (?,?,?)" etc is not a great idea.

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/models/dag.py (hunk 3 above)
Comment (ephraimbuddy) [reply airflow/models/dag.py]: Actually the cause of the issue was that an active dag had errors after upgrade, normally it should have max_active_runs but it didn’t have it because it failed to parse. 

I’ll update it to only include this check and remove import errors

File: airflow/models/dag.py (hunk 4 above)
Comment (ephraimbuddy) [reply airflow/models/dag.py]: I have updated it to exclude those with null max_active_runs.

**Comment by ashb** [comment]
This feels like only a partial fix.

If a DAG is currently working (and has max_active_runs) and _then_ the file is updated and now has an error, the DAG will still be scheduled, but it probably shouldn't be.

To do that without needing to join against ImportError perhaps we need to add a new column to DAG -- we have `is_active` which is close, but setting that to false hides it from the UI which isn't the behaviour we want here.

**Comment by ephraimbuddy** [comment]
> This feels like only a partial fix.
> 
> If a DAG is currently working (and has max_active_runs) and _then_ the file is updated and now has an error, the DAG will still be scheduled, but it probably shouldn't be.
> 
> To do that without needing to join against ImportError perhaps we need to add a new column to DAG -- we have `is_active` which is close, but setting that to false hides it from the UI which isn't the behaviour we want here.

I have updated it and it looks good. Thanks for the review!!

File: tests/models/test_dag.py [hunk 5]
```diff
@@ -1848,9 +1888,6 @@ def test_dags_needing_dagruns_not_too_early(self):
         dag_models = DagModel.dags_needing_dagruns(session).all()
         assert dag_models == []
 
-        session.rollback()
-        session.close()
```
Comment (ephraimbuddy) [inline tests/models/test_dag.py]: Removed this because of the added clean_db method

File: airflow/dag_processing/processor.py [hunk 6]
```diff
@@ -533,6 +533,10 @@ def update_import_errors(session: Session, dagbag: DagBag) -> None:
 
         # Add the errors of the processed files
         for filename, stacktrace in dagbag.import_errors.items():
+            dag_model = session.query(DagModel).filter(DagModel.fileloc == filename).first()
```
Comment (jedcunningham) [inline airflow/dag_processing/processor.py]: I don't think this can be on just the first DagModel, don't we need to set it on all of them from that file?

Comment (kaxil) [reply airflow/dag_processing/processor.py]:
```suggestion
            session.query(DagModel).filter(DagModel.fileloc == filename).update({"has_import_errors": True})
```

might be better _I think_

File: airflow/models/dag.py [hunk 7]
```diff
@@ -2880,11 +2884,14 @@ def dags_needing_dagruns(cls, session: Session):
         # TODO[HA]: Bake this query, it is run _A lot_
         # We limit so that _one_ scheduler doesn't try to do all the creation
         # of dag runs
+
         query = (
             session.query(cls)
             .filter(
+                cls.max_active_runs != None,  # noqa
```
Comment (ashb) [inline airflow/models/dag.py]:
```suggestion
```

We don't need this condition now I think, right?

File: docs/apache-airflow/migrations-ref.rst [hunk 8]
```diff
@@ -23,7 +23,9 @@ Here's the list of all the Database Migrations that are executed via when you ru
 +--------------------------------+------------------+-----------------+---------------------------------------------------------------------------------------+
 | Revision ID                    | Revises ID       | Airflow Version | Description                                                                           |
 +--------------------------------+------------------+-----------------+---------------------------------------------------------------------------------------+
-| ``7b2661a43ba3`` (head)        | ``142555e44c17`` | ``2.2.0``       | Change ``TaskInstance`` and ``TaskReschedule`` tables from execution_date to run_id.  |
+| ``be2bfac3da23`` (head)        | ``7b2661a43ba3`` | ``2.2.2``       | Add has_import_errors column to DagModel                                              |
```
Comment (ashb) [inline docs/apache-airflow/migrations-ref.rst]:
```suggestion
| ``be2bfac3da23`` (head)        | ``7b2661a43ba3`` | ``2.2.3``       | Add has_import_errors column to DagModel                                              |
```

File: tests/models/test_dag.py (hunk 5 above)
Comment (ashb) [reply tests/models/test_dag.py]: But with these rollbacks it _shouldn't_ need a clean method.

File: tests/models/test_dag.py [hunk 9]
```diff
@@ -1863,13 +1900,29 @@ def test_max_active_runs_not_none(self):
...
+            next_dagrun=None,
+            next_dagrun_create_after=None,
+            is_active=True,
+        )
+        # assert has_import_error is false
+        assert not orm_dag.has_import_errors
+        session.add(orm_dag)
+        session.flush()
+
+        assert not orm_dag.has_import_errors
```
Comment (ashb) [inline tests/models/test_dag.py]: This test isn't really adding much -- it's testing `self.has_import_errors = False` in the constructor.

It certainly doesn't need a whole separate test case for this, it can be added as an assert somewhere else if we want it. You already test basically everything here in `test_dags_needing_dagruns_doesnot_send_dagmodel_with_import_errors`, so this whole test function isn't needed

File: tests/models/test_dag.py [hunk 10]
```diff
@@ -1898,8 +1951,65 @@ def test_dags_needing_dagruns_only_unpaused(self):
...
+            next_dagrun_create_after=DEFAULT_DATE + timedelta(days=1),
+            is_active=True,
+        )
+        assert not orm_dag.has_import_errors
+        session.add(orm_dag)
+        session.flush()
+
+        needed = DagModel.dags_needing_dagruns(session).all()
+        assert needed == [orm_dag]
+        # set to None and ensure it's not sent to scheduler
```
Comment (ashb) [inline tests/models/test_dag.py]:
```suggestion
```

File: tests/models/test_dag.py [hunk 11]
```diff
@@ -1898,8 +1951,65 @@ def test_dags_needing_dagruns_only_unpaused(self):
...
+        orm_dag.max_active_runs = None
+        session.merge(orm_dag)
+        session.flush()
+        needed = DagModel.dags_needing_dagruns(session).all()
+        assert needed == []
+
```
Comment (ashb) [inline tests/models/test_dag.py]:
```suggestion
```

I don't think we need this behaviour _and_ the explicit flag, so since we have added the flag we should just use that.

In which case we should also re-title this PR please.

**Review by ashb — DISMISSED** [review]
Happy to discuss about if we still need the max_active_runs condition or not too.

File: airflow/models/dag.py (hunk 7 above)
Comment (ephraimbuddy) [reply airflow/models/dag.py]: Yes. Will remove it

File: tests/models/test_dag.py (hunk 5 above)
Comment (ephraimbuddy) [reply tests/models/test_dag.py]: Yes. I was seeing that it was repeated, happy to return it back if it's preferred

**Comment by ephraimbuddy** [comment]
> Happy to discuss about if we still need the max_active_runs condition or not too.

I agree we no longer need it.

I restricted the creation of dagruns on the UI and API too. Let me know if I should remove these restrictions

**Comment by ashb** [comment]
> > Happy to discuss about if we still need the max_active_runs condition or not too.
> 
> I agree we no longer need it.
> 
> I restricted the creation of dagruns on the UI and API too. Let me know if I should remove these restrictions

If you want to reduce duplication, but keep the rollback you can go from this:

```python
     def test_dags_needing_dagruns_not_too_early(self):
         ...
        session = settings.Session()
```
to this:

```python
     def test_dags_needing_dagruns_not_too_early(self, session):
```

And remove the `session = settings.Session()`. That session pytest fixture defined in `tests/conftest.py` already does the rollback.

But that should probably be a separate change to this PR.

**Comment by ephraimbuddy** [comment]
> > > Happy to discuss about if we still need the max_active_runs condition or not too.
> > 
> > 
> > I agree we no longer need it.
> > I restricted the creation of dagruns on the UI and API too. Let me know if I should remove these restrictions
> 
> If you want to reduce duplication, but keep the rollback you can go from this:
> 
> ```python
>      def test_dags_needing_dagruns_not_too_early(self):
>          ...
>         session = settings.Session()
> ```
> 
> to this:
> 
> ```python
>      def test_dags_needing_dagruns_not_too_early(self, session):
> ```
> 
> And remove the `session = settings.Session()`. That session pytest fixture defined in `tests/conftest.py` already does the rollback.
> 
> But that should probably be a separate change to this PR.

That's true, will work on it

File: docs/apache-airflow/migrations-ref.rst [hunk 12]
```diff
@@ -23,7 +23,9 @@ Here's the list of all the Database Migrations that are executed via when you ru
 +--------------------------------+------------------+-----------------+---------------------------------------------------------------------------------------+
 | Revision ID                    | Revises ID       | Airflow Version | Description                                                                           |
 +--------------------------------+------------------+-----------------+---------------------------------------------------------------------------------------+
-| ``7b2661a43ba3`` (head)        | ``142555e44c17`` | ``2.2.0``       | Change ``TaskInstance`` and ``TaskReschedule`` tables from execution_date to run_id.  |
+| ``be2bfac3da23`` (head)        | ``7b2661a43ba3`` | ``2.2.3``       | Add has_import_errors column to DagModel                                              |
```
Comment (ashb) [inline docs/apache-airflow/migrations-ref.rst:26]: I'm tempted to say this should be 2.3.0 -- it's maybe a bug fix, but this behaviour has been the same for all of 2.x so far.

What do others think here?

**Review by ashb — COMMENTED** [review]
LGTM now, just the 2.2.3 vs 2.3 decision.

File: docs/apache-airflow/migrations-ref.rst (hunk 12 above)
Comment (potiuk) [reply docs/apache-airflow/migrations-ref.rst:26]: I am for adding it to 2.2.3. There is hardly a use of DagRuns for DAGs with import errors, especially that you could not see whether there were errors via API. Even if it requires data model change, that's fine as patch-level. This is OK for a bugfix to require new field in the DB.

Comment (jedcunningham) [reply docs/apache-airflow/migrations-ref.rst:26]: Since this issue will crash the scheduler, I lean toward `2.2.3`.

File: tests/dag_processing/test_processor.py [hunk 13]
```diff
@@ -481,6 +481,34 @@ def test_add_unparseable_zip_file_creates_import_error(self, tmpdir):
...
+        with open(temp_dagfile, 'a') as file:
+            file.writelines(UNPARSEABLE_DAG_FILE_CONTENTS)
+
+        self._process_file(temp_dagfile, session)
+        import_errors = session.query(errors.ImportError).all()
+
+        assert len(import_errors) == 1
+        import_error = import_errors[0]
+        assert import_error.filename == temp_dagfile
+        assert import_error.stacktrace == f"invalid syntax ({TEMP_DAG_FILENAME}, line 53)"
```
Comment (jedcunningham) [inline tests/dag_processing/test_processor.py]:
```suggestion
```

nit: Do we need to assert the traceback?

File: tests/dag_processing/test_processor.py [hunk 14]
```diff
@@ -481,6 +481,34 @@ def test_add_unparseable_zip_file_creates_import_error(self, tmpdir):
...
+        session.rollback()
```
Comment (jedcunningham) [inline tests/dag_processing/test_processor.py]: I don't think this is actually doing anything - `update_import_errors` does a commit already. That said, this should probably stay and happen at the end of this test case.

File: tests/models/test_dag.py [hunk 15]
```diff
@@ -822,6 +822,36 @@ def test_bulk_write_to_db_max_active_runs(self, state):
...
 
+    def test_bulk_write_to_db_has_import_error(self):
+        """
+        Test that DagModel.has_import_error is set to false if no import errors.
+        """
+        dag = DAG(dag_id='test_has_import_error', start_date=DEFAULT_DATE)
+
+        DummyOperator(task_id='dummy', dag=dag, owner='airflow')
+
+        session = settings.Session()
```
Comment (jedcunningham) [inline tests/models/test_dag.py:833]:
```suggestion
    def test_bulk_write_to_db_has_import_error(self, session):
        """
        Test that DagModel.has_import_error is set to false if no import errors.
        """
        dag = DAG(dag_id='test_has_import_error', start_date=DEFAULT_DATE)

        DummyOperator(task_id='dummy', dag=dag, owner='airflow')

```

Use the session fixture?

File: tests/models/test_dag.py [hunk 16]
```diff
@@ -1901,6 +1932,36 @@ def test_dags_needing_dagruns_only_unpaused(self):
...
+        needed = DagModel.dags_needing_dagruns(session).all()
+        assert needed == [orm_dag]
+        orm_dag.has_import_errors = True
+        session.merge(orm_dag)
+        session.flush()
+        needed = DagModel.dags_needing_dagruns(session).all()
+        assert needed == []
+
+        session.rollback()
+        session.close()
```
Comment (jedcunningham) [inline tests/models/test_dag.py]:
```suggestion
```

The session fixture already does a rollback.

File: tests/models/test_dag.py (hunk 15 above)
Comment (ephraimbuddy) [reply tests/models/test_dag.py:833]: Ah, yeah. That's better

File: tests/dag_processing/test_processor.py (hunk 13 above)
Comment (ephraimbuddy) [reply tests/dag_processing/test_processor.py]: I think there's no harm in asserting it?

Comment (jedcunningham) [reply tests/dag_processing/test_processor.py]: At least in my view, it's needlessly fragile. since we append, that line number could change if that test dag is updated.

File: tests/models/test_dag.py (hunk 15 above)
Comment (ephraimbuddy) [reply tests/models/test_dag.py:833]: Turns out the class TestDag is using unittest.TestCase and not pytest. Wondering if it's worth it to convert it to pytest in this PR so we can use the session fixture or create a session fixture inside the unittest class as shown here?: https://pytest.org/en/latest/how-to/unittest.html#using-autouse-fixtures-and-accessing-other-fixtures

Comment (jedcunningham) [reply tests/models/test_dag.py:833]: I'd say do that in another PR. Plenty of other tests could benefit from it as well.

**Review by jedcunningham — APPROVED** [review]

**Comment by uranusjr** [comment]
Do we really need a new field? I think we can instead join `ImportError` on `import_error.filename = dag.fileloc` and error if the result is not empty.

**Comment by ephraimbuddy** [comment]
> Do we really need a new field? I think we can instead join `ImportError` on `import_error.filename = dag.fileloc` and error if the result is not empty.

It looks like it'll not be that efficient. See 
https://github.com/apache/airflow/pull/19367#discussion_r742659192

Not sure the impact the join would have.
cc: @ashb

**Comment by ashb** [comment]
Yeah, we _could_ do it with a join, but this is in the hot path for the main scheduler loop so it seems better to denormalise it here.

**Comment by uranusjr** [comment]
Denormalise to optimise makes sense.

**Review by uranusjr — APPROVED** [review]
